internal/engine/exit/controller.go

```diff
@@ -2,124 +2,331 @@ package exit
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	tiltruns "github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
+	"github.com/tilt-dev/tilt/pkg/logger"
 
 	v1 "k8s.io/api/core/v1"
 
 	"github.com/tilt-dev/tilt/internal/store"
 	"github.com/tilt-dev/tilt/pkg/model"
 )
 
-// Controls normal process termination. Either Tilt completed all its work,
+// Controller handles normal process termination. Either Tilt completed all its work,
 // or it determined that it was unable to complete the work it was assigned.
 type Controller struct {
+	pid       int64
+	startTime time.Time
+	client    ctrlclient.Client
+	tiltRun   *tiltruns.TiltRun
 }
 
-func NewController() *Controller {
-	return &Controller{}
+func (c *Controller) initialize(ctx context.Context, state store.EngineState) error {
+	c.tiltRun = &tiltruns.TiltRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "Tiltfile",
+		},
+		Spec: tiltruns.TiltRunSpec{
+			TiltfilePath: state.TiltfilePath,
+		},
+		Status: tiltruns.TiltRunStatus{
+			PID:       c.pid,
+			StartTime: metav1.NewMicroTime(c.startTime),
+		},
+	}
+
+	switch state.EngineMode {
+	case store.EngineModeUp:
+		c.tiltRun.Spec.ExitCondition = tiltruns.ExitConditionManual
+	case store.EngineModeCI:
+		c.tiltRun.Spec.ExitCondition = tiltruns.ExitConditionFirstFailure
+	}
+
+	if err := c.client.Create(ctx, c.tiltRun); err != nil {
```

Comment:
i've been wondering if a good rule of thumb is that controllers should be idempotent (so that Create() on a resource that already exists should be OK)

Reply:
I agree (and reconciler pattern helps with that a lot, since you can basically just requeue on a lot of things and take care of it the next pass).

In this case, it'd mean extra logic to handle things (e.g. fetch if already created, similar-ish logic on any conflict on update failure) - unless you feel strongly, I _think_ that since this is the only think writing to it at the moment, it'd be better to just let it always assume these operations shouldn't fail.

(The `Create` is actually a really weird one - I added a `TODO` because I think it should get created _explicitly_ as part of Tiltfile/configs reloading rather than implicitly here, and then this can behave much closer to a real reconciler and just always read from cached client, at which point it _could_ trivially handle optimistic concurrency failures during update.)